DOC-4363: Added member reregistration steps and updated member endpoints section (#1730)
corda/corda-docs-portal: content/en/platform/corda/5.1/application-networks/creating/members/reregister.md

---
date: '2023-10-04'
version: 'Corda 5.1'
title: "Re-register a Member"
menu:
  corda51:
    identifier: corda51-networks-members-reregister
    parent: corda51-networks-members
    weight: 6000
section_menu: corda51
---

# Re-register a Member

A {{< tooltip >}}member{{< /tooltip >}} may request to update its own member-provided context, for example,
after key rotation or changes to its endpoint information. Additionally, a member who previously attempted to register but
failed may wish to try again. The membership re-registration steps described in this section can be followed for both of these scenarios.

The instructions on this page assume that you have completed the [registration]({{< relref "register.md" >}}) steps.

You can learn more about configuring the registration process in the [Managing Members section]({{< relref "../../managing/registration-requests/_index.md" >}}).

1. Inspect the member-provided context.

   Currently, updates to the member-provided context are limited to custom properties (keys with the `ext.` prefix) and endpoint
   information only. Changes to other Corda platform properties are not currently supported.

>    You may inspect member's current member-provided context either by performing a member lookup, or by looking up its latest

Reviewer: ```suggestion
   You can inspect a member's current member-provided context either by performing a member lookup, or by looking up its latest
```

     "corda.ledger.keys.0.id": "'$LEDGER_KEY_ID'",
     "corda.ledger.keys.0.signature.spec": "SHA256withECDSA",
     "corda.endpoints.0.connectionURL": "https://'$P2P_GATEWAY_HOST':'$P2P_GATEWAY_PORT'",
     "corda.endpoints.0.protocolVersion": "1",
     "ext.sample": "apple"
   }'
   ```
   {{% /tab %}}
   {{% tab name="PowerShell" %}}
   ```shell
   $REGISTRATION_CONTEXT = @{
     'corda.session.keys.0.id' =  $SESSION_KEY_ID
     'corda.session.keys.0.signature.spec' = "SHA256withECDSA"
     'corda.ledger.keys.0.id' = $LEDGER_KEY_ID
     'corda.ledger.keys.0.signature.spec' = "SHA256withECDSA"
     'corda.endpoints.0.connectionURL' = "https://$P2P_GATEWAY_HOST`:$P2P_GATEWAY_PORT"
     'corda.endpoints.0.protocolVersion' = "1",
     'ext.sample' = "apple"
   }
   ```
   {{% /tab %}}
   {{< /tabs >}}

3. Optional: Include serial number in the registration context.

   The `corda.serial` Corda platform property is embedded in the `MemberInfo` of all members. It acts as the `MemberInfo`'s version,
   and is incremented by 1 after each registration.

   This serial number may be optionally included in the registration context to specify which version of the `MemberInfo` was

>    intended to be updated. If no serial number is provided while re-registering, the platform will use the member's latest

Reviewer: ```suggestion
   intended to be updated. If no serial number is provided while re-registering, the platform uses the member's latest
```

   {{< note >}}
   For first-time registration the serial number should be 0, if provided.
   {{< /note >}}

   {{< tabs >}}
   {{% tab name="Bash"%}}
   ```bash
   export REGISTRATION_CONTEXT='{
     "corda.session.keys.0.id": "'$SESSION_KEY_ID'",
     "corda.session.keys.0.signature.spec": "SHA256withECDSA",
     "corda.ledger.keys.0.id": "'$LEDGER_KEY_ID'",
     "corda.ledger.keys.0.signature.spec": "SHA256withECDSA",
     "corda.endpoints.0.connectionURL": "https://'$P2P_GATEWAY_HOST':'$P2P_GATEWAY_PORT'",
     "corda.endpoints.0.protocolVersion": "1",
     "ext.sample": "apple",
     "corda.serial": "1"
   }'
   ```
   {{% /tab %}}
   {{% tab name="PowerShell" %}}
   ```shell
   $REGISTRATION_CONTEXT = @{
     'corda.session.keys.0.id' =  $SESSION_KEY_ID
     'corda.session.keys.0.signature.spec' = "SHA256withECDSA"
     'corda.ledger.keys.0.id' = $LEDGER_KEY_ID
     'corda.ledger.keys.0.signature.spec' = "SHA256withECDSA"
     'corda.endpoints.0.connectionURL' = "https://$P2P_GATEWAY_HOST`:$P2P_GATEWAY_PORT"
     'corda.endpoints.0.protocolVersion' = "1",
     'ext.sample' = "apple",
     'corda.serial' = "1"
   }
   ```
   {{% /tab %}}
   {{< /tabs >}}

> 4. Send a re-registration request using the common registration/re-registration endpoint.

Reviewer: Might be nice to name the endpoint and/or link to it in the API docs.

Author: @YashNabar Is that the correct endpoint? https://docs.r3.com/en/platform/corda/5.0/reference/rest-api/openapi.html#tag/Member-Registration-API/operation/post_membership__holdingidentityshorthash_

Reviewer 2: Yes, that is correct @AnnaK-R3

registration request. For example, to look up Alice:

{{< tabs >}}
{{% tab name="Bash"%}}
```bash
curl --insecure -u admin:admin -X GET $API_URL/members/$HOLDING_ID?O=Alice
```
{{% /tab %}}
{{% tab name="PowerShell" %}}
```shell
Invoke-RestMethod -SkipCertificateCheck  -Headers @{Authorization=("Basic {0}" -f $AUTH_INFO)} -Uri "$API_URL/membership/$HOLDING_ID?O=Alice" | ConvertTo-Json -Depth 4
```
{{% /tab %}}
{{< /tabs >}}

Alternatively, to retrieve the registration request:

{{< tabs >}}
{{% tab name="Bash"%}}
```bash
export REGISTRATION_ID=<registration ID>
curl --insecure -u admin:admin -X GET $API_URL/membership/$HOLDING_ID/$REGISTRATION_ID
```
{{% /tab %}}
{{% tab name="PowerShell" %}}
```shell
$REGISTRATION_ID = <registration ID>
Invoke-RestMethod -SkipCertificateCheck  -Headers @{Authorization=("Basic {0}" -f $AUTH_INFO)} -Uri "$API_URL/membership/$HOLDING_ID/$REGISTRATION_ID"
```
{{% /tab %}}
{{< /tabs >}}

To retrieve information about key pairs (for example, key ID required for the registration context) belonging to a
holding identity, use:

```bash
curl --insecure -u admin:admin -X GET $API_URL/keys/$HOLDING_ID
```

> ## Re-register a Member

Reviewer: Seems strange to have a sub-section named the same as a parent. Might suggest you don't need subsections at all?

Author: Ok, changed it to a numbered step. My only fear was that it would become very long but I guess indentation helps to make things more structured and legible.

> 1. Get a member-provided context of a member who has previously registered successfully, for example:

Reviewer: ```suggestion
1. Retrieve a member-provided context of a member who has previously registered successfully, for example:
```

> {{< tabs >}}

Reviewer: Code block should be indented in line with the step.

Author: indented

{{% tab name="Bash"%}}
```bash
REGISTRATION_CONTEXT='{
  "corda.session.keys.0.id": "'$SESSION_KEY_ID'",
  "corda.session.keys.0.signature.spec": "SHA256withECDSA",
  "corda.ledger.keys.0.id": "'$LEDGER_KEY_ID'",
  "corda.ledger.keys.0.signature.spec": "SHA256withECDSA",
  "corda.endpoints.0.connectionURL": "https://'$P2P_GATEWAY_HOST':'$P2P_GATEWAY_PORT'",
  "corda.endpoints.0.protocolVersion": "1"
}'
```
{{% /tab %}}
{{% tab name="PowerShell" %}}
```shell
$REGISTRATION_CONTEXT = @{
  'corda.session.keys.0.id' =  $SESSION_KEY_ID
  'corda.session.keys.0.signature.spec' = "SHA256withECDSA"
  'corda.ledger.keys.0.id' = $LEDGER_KEY_ID
  'corda.ledger.keys.0.signature.spec' = "SHA256withECDSA"
  'corda.endpoints.0.connectionURL' = "https://$P2P_GATEWAY_HOST`:$P2P_GATEWAY_PORT"
  'corda.endpoints.0.protocolVersion' = "1"
}
```
{{% /tab %}}
{{< /tabs >}}

2. Add a custom property to the member-provided context. In this example, the updated context contains the new custom property with the `ext.` prefix:

> {{< tabs >}}

Reviewer: Code block should be indented in line with the step.

Author: indented

{{% tab name="Bash"%}}
```bash
export REGISTRATION_CONTEXT='{
  "corda.session.keys.0.id": "'$SESSION_KEY_ID'",
  "corda.session.keys.0.signature.spec": "SHA256withECDSA",
  "corda.ledger.keys.0.id": "'$LEDGER_KEY_ID'",
  "corda.ledger.keys.0.signature.spec": "SHA256withECDSA",
  "corda.endpoints.0.connectionURL": "https://'$P2P_GATEWAY_HOST':'$P2P_GATEWAY_PORT'",
  "corda.endpoints.0.protocolVersion": "1",
  "ext.sample": "apple"
}'
```
{{% /tab %}}
{{% tab name="PowerShell" %}}
```shell
$REGISTRATION_CONTEXT = @{
  'corda.session.keys.0.id' =  $SESSION_KEY_ID
  'corda.session.keys.0.signature.spec' = "SHA256withECDSA"
  'corda.ledger.keys.0.id' = $LEDGER_KEY_ID
  'corda.ledger.keys.0.signature.spec' = "SHA256withECDSA"
  'corda.endpoints.0.connectionURL' = "https://$P2P_GATEWAY_HOST`:$P2P_GATEWAY_PORT"
  'corda.endpoints.0.protocolVersion' = "1",
  'ext.sample' = "apple"
}
```
{{% /tab %}}
{{< /tabs >}}

> 3. Optional: Include serial number in the registration context.

Reviewer: The content that follows this should also be indented inline with the step.

Author: indented

registration request. For example, to look up Alice:

{{< tabs >}}
{{% tab name="Bash"%}}
```bash
curl --insecure -u admin:admin -X GET $API_URL/members/$HOLDING_ID?O=Alice
```
{{% /tab %}}
{{% tab name="PowerShell" %}}
```shell
Invoke-RestMethod -SkipCertificateCheck  -Headers @{Authorization=("Basic {0}" -f $AUTH_INFO)} -Uri "$API_URL/membership/$HOLDING_ID?O=Alice" | ConvertTo-Json -Depth 4
```
{{% /tab %}}
{{< /tabs >}}

Alternatively, to retrieve the registration request:

{{< tabs >}}
{{% tab name="Bash"%}}
```bash
export REGISTRATION_ID=<registration ID>
curl --insecure -u admin:admin -X GET $API_URL/membership/$HOLDING_ID/$REGISTRATION_ID
```
{{% /tab %}}
{{% tab name="PowerShell" %}}
```shell
$REGISTRATION_ID = <registration ID>
Invoke-RestMethod -SkipCertificateCheck  -Headers @{Authorization=("Basic {0}" -f $AUTH_INFO)} -Uri "$API_URL/membership/$HOLDING_ID/$REGISTRATION_ID"
```
{{% /tab %}}
{{< /tabs >}}

To retrieve information about key pairs (for example, key ID required for the registration context) belonging to a
holding identity, use:

```bash
curl --insecure -u admin:admin -X GET $API_URL/keys/$HOLDING_ID
```

> ## Re-register a Member

Reviewer: The content in this section seems like it should be a set of tasks?

Author: @nadinequinr3 What do you mean by "a set of tasks"? Numbered steps?

using the common registration/re-registration endpoint:

{{< tabs >}}
{{% tab name="Bash"%}}
```bash
export REGISTRATION_REQUEST='{"memberRegistrationRequest":{"context": '$REGISTRATION_CONTEXT'}}'
curl --insecure -u admin:admin -d "$REGISTRATION_REQUEST" $API_URL/membership/$HOLDING_ID
```
{{% /tab %}}
{{% tab name="PowerShell" %}}
```shell
$RESGISTER_RESPONSE = Invoke-RestMethod -SkipCertificateCheck  -Headers @{Authorization=("Basic {0}" -f $AUTH_INFO)} -Method Post -Uri "$API_URL/membership/$HOLDING_ID" -Body (ConvertTo-Json -Depth 4 @{
    memberRegistrationRequest = @{
        context = $REGISTRATION_CONTEXT
    }
})
$RESGISTER_RESPONSE.registrationStatus
```
{{% /tab %}}
{{< /tabs >}}

This sends the request to the MGM, which should return a successful response with status `SUBMITTED`. You can check if the
request was approved by checking the status of the registration request.

After successful re-registration, you should be able to see the member's information containing the `ext.sample` field in
their member-provided context. The member retains its most recent status in the group, for example, a suspended member
will remain suspended after successful re-registration.

> ## **Optional:** Include Serial Number in the Registration Context

Author: Updated

You can retrieve the serial number from the MGM-provided context of the `MemberInfo` by performing a member lookup.

{{< note >}}
For first-time registration the serial number should be 0, if provided.
{{< /note >}}

{{< tabs >}}
{{% tab name="Bash"%}}
```bash
export REGISTRATION_CONTEXT='{
  "corda.session.keys.0.id": "'$SESSION_KEY_ID'",
  "corda.session.keys.0.signature.spec": "SHA256withECDSA",
  "corda.ledger.keys.0.id": "'$LEDGER_KEY_ID'",
  "corda.ledger.keys.0.signature.spec": "SHA256withECDSA",
  "corda.endpoints.0.connectionURL": "https://'$P2P_GATEWAY_HOST':'$P2P_GATEWAY_PORT'",
  "corda.endpoints.0.protocolVersion": "1",
  "ext.sample": "apple",
  "corda.serial": "1"
}'
```
{{% /tab %}}
{{% tab name="PowerShell" %}}
```shell
$REGISTRATION_CONTEXT = @{
  'corda.session.keys.0.id' =  $SESSION_KEY_ID
  'corda.session.keys.0.signature.spec' = "SHA256withECDSA"
  'corda.ledger.keys.0.id' = $LEDGER_KEY_ID
  'corda.ledger.keys.0.signature.spec' = "SHA256withECDSA"
  'corda.endpoints.0.connectionURL' = "https://$P2P_GATEWAY_HOST`:$P2P_GATEWAY_PORT"
  'corda.endpoints.0.protocolVersion' = "1",
  'ext.sample' = "apple",
  'corda.serial' = "1"
}
```
{{% /tab %}}
{{< /tabs >}}

> ## Request Queue

Reviewer: Again, I'm not sure this needs to be a whole subsection?

Author: Updated

> If a member submits more than one registration request at the same time, the MGM will queue the requests and process them

Reviewer: ```suggestion
If a member submits more than one registration request at the same time, the MGM queuea the requests and processes them
```

one by one, treating each subsequent request in the queue as a re-registration attempt.

> If a member submits multiple re-registration requests with the same serial number, the first request will be processed

Reviewer: ```suggestion
If a member submits multiple re-registration requests with the same serial number, the MGM processes the first request
```

> (if the serial number is valid). However, the other requests will be declined because the serial number specified in

Reviewer: ```suggestion
(if the serial number is valid). The MGM declines the other requests because the serial number specified in
```

> those requests would be outdated after the first request is completed.

Reviewer: ```suggestion
those requests is outdated after the first request is completed.
```

  "corda.endpoints.0.protocolVersion": "1",
  "ext.sample": "apple",
  "corda.serial": "1"
}'
```
{{% /tab %}}
{{% tab name="PowerShell" %}}
```shell
$REGISTRATION_CONTEXT = @{
  'corda.session.keys.0.id' =  $SESSION_KEY_ID
  'corda.session.keys.0.signature.spec' = "SHA256withECDSA"
  'corda.ledger.keys.0.id' = $LEDGER_KEY_ID
  'corda.ledger.keys.0.signature.spec' = "SHA256withECDSA"
  'corda.endpoints.0.connectionURL' = "https://$P2P_GATEWAY_HOST`:$P2P_GATEWAY_PORT"
  'corda.endpoints.0.protocolVersion' = "1",
  'ext.sample' = "apple",
  'corda.serial' = "1"
}
```
{{% /tab %}}
{{< /tabs >}}

## Request Queue

If a member submits more than one registration request at the same time, the MGM will queue the requests and process them
one by one, treating each subsequent request in the queue as a re-registration attempt.

If a member submits multiple re-registration requests with the same serial number, the first request will be processed
(if the serial number is valid). However, the other requests will be declined because the serial number specified in
those requests would be outdated after the first request is completed.

> ## MGM Admin API

Reviewer 2: This section should probably be under Managing Members, as this API is for the MGM operator

Author: Moved

   {{% /tab %}}
   {{< /tabs >}}

> 4. Send a re-registration request using the common [registration/re-registration endpoint](https://docs.r3.com/en/platform/corda/5.0/reference/rest-api/openapi.html#tag/Member-Registration-API/operation/post_membership__holdingidentityshorthash_).

Reviewer: I think you should be able to specify this as a relative link.. probably ../../../reference/rest-api/openapi.html#tag/Member-Registration-API/operation/post_membership__holdingidentityshorthash_

Author: updated
